livre/core/cache/CacheObject.h

```diff
@@ -91,88 +93,78 @@ class CacheObject
     LIVRECORE_API void setUnloadable( bool unloadable );
 
     /**
-     * getReferenceCount_ Should not be called by user. The function is threadsafe.
      * @return The number of references to CacheObject.
      */
-    LIVRECORE_API uint32_t getReferenceCount_() const;
-
-    /**
-     * updateLastUsed_ Should not be called by user.
-     * @param lastUsedTime Update the last used time.
-     */
-    void updateLastUsed_( const double lastUsedTime );
-
-    /**
-     * updateLastUsedWithCurrentTime_ Updates the last used time with current time.
-     */
-    LIVRECORE_API void updateLastUsedWithCurrentTime_();
-
-    /**
-     * resetLastUsed_ Resets the last used time to 0.
-     */
-    void resetLastUsed_();
-
-    LIVRECORE_API virtual ~CacheObject() { }
+    LIVRECORE_API uint32_t getRefCount() const;
 
     /**
-     * Register an observer for operations done on the object.
-     * @param observer The observer to add.
+     * Updates the last used time with current time.
      */
-    LIVRECORE_API void registerObserver( CacheObjectObserver* observer );
+    LIVRECORE_API void touch();
 
     /**
-     * Unregisters an observer from the object
-     * @param observer The observer to remove.
+     * @return On default returns true if cache ids are same
      */
-    LIVRECORE_API void unregisterObserver( CacheObjectObserver* observer );
+    virtual bool operator==( const CacheObject& cacheObject ) const;
 
 protected:
 
     friend class Cache;
     friend class CacheStatistics;
 
-    LIVRECORE_API CacheObject();
+    LIVRECORE_API CacheObject( const CacheId& cacheId = INVALID_CACHE_ID );
 
     /**
      * Implemented by the derived class, for loading data to memory. Thread safety is satisfied for
      * loading and unloading.
      * @return True if data is loaded.
      */
-    virtual bool load_() = 0;
+    virtual bool _load() = 0;
 
     /**
      * Implemented by the derived class, for unloading data from memory. Thread safety is satisfied for
      * loading and unloading.
      */
-    virtual void unload_() = 0;
+    virtual void _unload() = 0;
 
     /**
      * @return The validity of derived class.
      */
-    virtual bool isValid_() const = 0;
+    virtual bool _isValid() const;
 
     /**
      * @return True if object is loaded into memory.
      */
-    virtual bool isLoaded_() const = 0;
+    virtual bool _isLoaded() const = 0;
 
     /**
      * @return The unconst object.
      */
-    CacheObject* getUnconst_() const { return const_cast< CacheObject * >( this ); }
+    CacheObject* _getUnconst() const { return const_cast< CacheObject * >( this ); }
```

Comment:
Either this way or  the _increaseReference() and decRef() methods should be consted and already the refcount is mutable. These methods are executing implicitly _unload() whenever refcount is zero then _unload() should be const too where it is a virtual method and doing a const virtual method enforces the derived object to have mutables inside. 

Whether the object is const or not in the they should increase, decrease references and should be unloaded when the refcount is 0. This is the idea so on this line it is correct but looks ugly. 

Or we never have const cache objects in intrusive_ptr s. 

Or ?